Content.Server/_ES/Masks/Masquerades/ESMasqueradeSystem.cs

```diff
@@ -0,0 +1,212 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Content.Server._ES.Masks.Masquerades;
+using Content.Server.Chat.Managers;
+using Content.Server.GameTicking;
+using Content.Server.GameTicking.Rules;
+using Content.Server.Mind;
+using Content.Shared._Citadel.Utilities;
+using Content.Shared._ES.Masks;
+using Content.Shared._ES.Masks.Components;
+using Content.Shared._ES.Masks.Masquerades;
+using Content.Shared.GameTicking.Components;
+using Content.Shared.Mind;
+using Content.Shared.Random.Helpers;
+using Robust.Shared.Player;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
+using Robust.Shared.Utility;
+
+namespace Content.Server._ES.Masks.Masquerades;
+
+/// <summary>
+///     This handles masquerade management and how they influence game flow.
+/// </summary>
+public sealed class ESMasqueradeSystem : GameRuleSystem<ESMasqueradeRuleComponent>
+{
+    [Dependency] private readonly IChatManager _chat = default!;
+    [Dependency] private readonly IPrototypeManager _proto = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly GameTicker _gameTicker = default!;
+    [Dependency] private readonly ESMaskSystem _mask = default!;
+    [Dependency] private readonly MindSystem _mind = default!;
+
+    // Icky global state.
+    private ProtoId<ESMasqueradePrototype>? _forcedMasquerade = null;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<AssignLatejoinerToTroupeEvent>(OnAssignLatejoiner);
+        SubscribeLocalEvent<AssignPlayersToTroupeEvent>(OnAssignPlayers);
+    }
+
+    private void OnAssignPlayers(ref AssignPlayersToTroupeEvent ev)
+    {
+        var rule = EntityQuery<ESMasqueradeRuleComponent>().SingleOrDefault();
+
+        if (rule is null)
+            return;
+
+        // TODO: This shouldn't be hardcoded here, TryGetMasks should be on MasqueradeKind.
+        if (rule.Masquerade!.Masquerade is not MasqueradeRoleSet set)
+            return;
```

Comment:
later pr todo or was this missed (not entirely sure what its implying)

Reply:
Later PR, I need to refactor all of the old logic and didn't want to do so this go.